src/ndarray/ndarray.cc

```diff
@@ -518,43 +515,57 @@ void CopyFromTo(const NDArray& from, const NDArray& to, int priority) {
   CHECK(from.shape().ndim() != 0)
       << "source operands have zero dimension shape";
   // important: callback must always capture by value
-  int a = from.ctx().dev_mask();
+  const auto from_ctx = from.ctx();
+  int a = from_ctx.dev_mask();
   int b = to.ctx().dev_mask();
   std::vector<Engine::VarHandle> const_vars;
   if (from.var() != to.var()) const_vars.push_back(from.var());
 
+  auto from_stype = from.storage_type();
+  auto to_stype = to.storage_type();
+
+  std::vector<Engine::VarHandle> mutable_vars(1, to.var());
+
+  std::vector<Resource> requested;
+  if (a == gpu::kDevMask && from_stype != to_stype) {
```

Comment:
What if b is on GPU ?

Reply:
Accordding to original codes,
```
-  std::vector<Resource> requested;
 -  if (is_same<from_xpu, mshadow::gpu>::value && from_stype != to_stype) {
 -    requested.push_back(ResourceManager::Get()->Request(from_ctx,
 -        ResourceRequest(ResourceRequest::kTempSpace)));
 -  }
```

Seems that whether temp space is used is irrelevant with the context of b ?

Reply:
Oh right. No need to request temp space if cast_storage happens on CPU.